src/map/script.c

```diff
@@ -6752,169 +6752,218 @@ static BUILDIN(warpchar)
-	int gid          = script_getnum(st, 5);
+	const char *m_name_to = script_getstr(st, 2);
+	const int type = (strcmp(m_name_to, "Random") == 0) ? 0 :
+			 (strcmp(m_name_to, "SavePointAll") == 0) ? 1 :
+			 (strcmp(m_name_to, "SavePoint") == 0) ? 2 : 3;
+	int map_index = 0;
 
-	if (script_hasdata(st, 6)) {
-		map_id = map->mapname2mapid(script_getstr(st, 6));
+	if (type == 3 && (map_index = script->mapindexname2id(st, m_name_to)) == 0) {
+		ShowError("script:%s: Target map not found! (%s)\n", script->getfuncname(st), m_name_to);
+		script_pushint(st, 0);
+		return false;
 	}
 
-	g = guild->search(gid);
-	if (g == NULL)
-		return true;
+	int x = script_getnum(st, 3);
+	int y = script_getnum(st, 4);
 
-	type = (strcmp(str, "Random") == 0) ? 0
-	: (strcmp(str, "SavePointAll") == 0) ? 1
-	: (strcmp(str, "SavePoint") == 0) ? 2
-	: 3;
+	if (type == 2) {
+		struct map_session_data *sd = script->rid2sd(st);
 
-	if (type == 2 && (sd = script->rid2sd(st)) == NULL)
-	{// "SavePoint" uses save point of the currently attached player
-		return true;
+		if (sd == NULL) {
+			ShowError("script:%s: No character attached for warp to save point!\n",
+				  script->getfuncname(st));
+			script_pushint(st, 0);
+			return false;
+		}
+
+		map_index = sd->status.save_point.map;
+		x = sd->status.save_point.x;
+		y = sd->status.save_point.y;
 	}
 
-	for (i = 0; i < MAX_GUILD; i++) {
-		if (g->member[i].online && g->member[i].sd != NULL) {
-			struct map_session_data *pl_sd = g->member[i].sd;
+	for (int i = 0; i < MAX_GUILD; i++) {
+		struct map_session_data *g_sd = g->member[i].sd;
 
-			if (map_id >= 0 && map_id != pl_sd->bl.m)
+		if (g->member[i].online == 0 || g_sd == NULL || g_sd->status.guild_id != g_id || pc_isdead(g_sd))
+			continue;
+
+		int offset = 0;
+		bool ignore_mapflags = false;
+
+		if (script_hasdata(st, 6) && script_isinttype(st, 6)) {
+			offset = 1;
+			ignore_mapflags = (script_getnum(st, 6) != 0);
+		}
+
+		if (!ignore_mapflags) {
+			if (((type == 0 || type > 2) && map->list[g_sd->bl.m].flag.nowarp == 1) ||
+			    (type > 0 && map->list[g_sd->bl.m].flag.noreturn == 1))
 				continue;
+		}
 
-			switch (type)
-			{
-			case 0: // Random
-				if (!map->list[pl_sd->bl.m].flag.nowarp)
-					pc->randomwarp(pl_sd, CLR_TELEPORT);
-				break;
-			case 1: // SavePointAll
-				if (!map->list[pl_sd->bl.m].flag.noreturn)
-					pc->setpos(pl_sd, pl_sd->status.save_point.map, pl_sd->status.save_point.x, pl_sd->status.save_point.y, CLR_TELEPORT);
-				break;
-			case 2: // SavePoint
-				if (!map->list[pl_sd->bl.m].flag.noreturn)
-					pc->setpos(pl_sd, sd->status.save_point.map, sd->status.save_point.x, sd->status.save_point.y, CLR_TELEPORT);
-				break;
-			case 3: // m,x,y
-				if (!map->list[pl_sd->bl.m].flag.noreturn && !map->list[pl_sd->bl.m].flag.nowarp)
-					pc->setpos(pl_sd, script->mapindexname2id(st, str), x, y, CLR_TELEPORT);
-				break;
-			}
+		const char *m_name_from = script_hasdata(st, 6 + offset) ? script_getstr(st, 6 + offset) : NULL;
```

Comment:
Just a thought, would it be nice to assign these variables outside the loop instead of reassigning it everytime in loop?

Reply:
Nope, I discussed this with @hemagx and it should be inside the loop.

Reply:
yes, look like calculation what not depend on ``i`` can be moved outside of ``for``. if need for safety can be added ``const`` to vars.

why move outside? because normally this code will be executed again and again for each ``for`` iteration, and this is can be slow.

Reply:
Unless you have a good reason to act different, variables should be declared in the smallest possible scope, which, in this case, is the `for` loop. This standard will be added to our style, when I'm ready to update it. (See topic in staff forum.)

<details><summary>Quote from my chat with @hemagx if you're interested:</summary>
 
    **Kenpachi**
    Assuming you need a variable only inside a for loop and it will have the same value for every iteration.
    Should it still be declared inside the loop?
 
    **Kenpachi**
    https://softwareengineering.stackexchange.com/a/296724
    This guy for example has good arguments to declare it outside the loop.
    https://stackoverflow.com/a/7959658
    But this guy has good arguments to declare it inside the loop.
    That's fucking confusing... :scream:
 
    **Ema**
    do you understand the Big O notation?
 
    **Kenpachi**
    I know the basics...
 
    **Ema**
    so think of it
    is it worth it or not
    99% it's not.
    and you should keep it inside loop no matter what the costs are
 
    **Kenpachi**
    I know the costs are low in my case, since the iteration count is very low.
    But it seems to be so badly unlogical/unneecessary to me, to set foo to n for x times. You know?
 
    **Ema**
    Safety comes before efficiency.
    show me the case tho?
 
    **Kenpachi**
    My "problem" is more like rating the benefit of keep the variable in the smallest possible scope.
 
    **Ema**
    the benefit?
    we have 10 years old bugs that no one was able to fix until we started fix scopes.
 
    **Kenpachi**
    Really? Well, that's a benefit of course. :rofl:
 
    **Ema**
    variables with no hard defined scopes are also prune to memory violations of all kinds << happens all the time.
 
    **Kenpachi**
    Okay, so like you said. "Safety comes before efficiency."
    That's an argument I can deal with. :slight_smile:
 
    **Ema**
    sometimes it can be the other way around
    you just need to have a really good argument.
    but as i said 99% it will never be good argument
   </details>

Reply:
this talk not about variables but about code what can be run only once.
inside for or while this code will be run N times. this is huge performance difference.

Example 1
```
int k;  // wrong
for (int i = 0; i < 10; i++)
{
  k = 10;
  // do something here with i and k
}
```
fixed version will be:

```
for (int i = 0; i < 10; i++)
{
  int k = 10;
  // do something here with k
}
```

Example 2

```
for (int i = 0; i < 10; i++)
{
  int k = call_some_slow_function();  // wrong
  // do something here with i and k
}
```
fixed version will be:


```
int k = call_some_slow_function();
for (int i = 0; i < 10; i++)
{
  // do something here with i and k
}
```

And in this example function ``call_some_slow_function`` not depend on ``i`` and this mean it can be moved outside of ``for``

Reply:
I actually do agree with 4144, it's slow function also const adds enough safety level.

Reply:
Changed...